pkg/lexer/lexer.go

```diff
@@ -1,5 +1,215 @@
 package lexer
 
+import (
+	"strings"
+	"unicode"
+)
+
 func Run(input string) string {
-	return "Lexing:\n" + input
+	l := New(input)
+	tok := l.Next()
+	var sb strings.Builder
+	for ; tok.Type != EOF; tok = l.Next() {
+		sb.WriteString(tok.String() + "\n")
+	}
+	sb.WriteString(tok.String() + "\n")
+	return sb.String()
+}
+
+type Lexer struct {
+	input []rune
+	r     rune // current rune under examination
+	pos   int  // current position in input (points to current rune)
+	line  int
+	col   int
+}
+
+func New(input string) *Lexer {
+	return &Lexer{input: []rune(input), pos: -1, line: 1}
+}
+
+func (l *Lexer) readRune() {
+	l.pos += 1
```

Comment:
What do you think of this for the read/peek rune functions?

```
func (l *Lexer) readRune() {
        if l.r == '\n' {
                l.line++
                l.col = 0
        }
        l.pos++
        l.col++
        l.r = l.lookAt(l.pos)
}

func (l *Lexer) peekRune() rune {
        return l.lookAt(l.pos + 1)
}

func (l *Lexer) peekRune2() rune {
        return l.lookAt(l.pos + 2)
}

func (l *Lexer) lookAt(pos int) rune {
        if pos >= len(l.input) {
                return 0
        }
        return l.input[pos]
}
```

Reply:
so good. thanks.

I've updated this. Do you still think this should go before `Next` and all the `readString|Num|Ident|Comment` functions after?

Reply:
Just thought about this a bit more and perhaps (not tested code, just written in the browser):
```
func (l *Lexer) skipWhitespace() {
    l.readWhile(func(r rune) bool { return r == ' ' || r == '\t' || r == '\r' })
```

Now, because skipWhitespace uses readWhile, it peeks rather than requiring `l.r` be read. This means the top of `Next()` can be written as:

````
func (l *Lexer) Next() *Token {
    l.skipWhitespace()
    l.readRune()

    ...
}
```

and I think this means you can now move the helper functions after `Next()`. This is because I found it weird that you would first read a rune, then skip whitespace, so I needed to know what they actually did. In the order of "skip whitespace" then "read rune", it just makes more sense I think - skip all the whitespace up to the next rune.

The only "oddity" now is that skipWhitespace does not skip newlines. Maybe (just maybe) skipWhitespace could be `skipHorizontalWhitespace`?